### PR TITLE
Handle dirty bits correctly when render objects are re-attached

### LINKED ISSUE #2855: Dirty state is not properly preserved when render objects are transplanted
Per discussion related to #2723:

When render objects are flagged as needing layout, compositing, painting, or semantics update, the logic for whether to add the object to the relevant dirty list is non-trivial. This logic needs to be invoked intelligently when attach() is called, so that re-parenting works properly.

There are existing cases where this is not handled properly: for example, if an object is painted in one frame (so that its layer is cached), then removed from the tree on the next frame, then marked as needing repaint, then attached again, the system will re-use the cached layer from the first frame rather than properly repainting on the next frame.

The initial commit for #2723 may exacerbate the situation because flagging an object as dirty can't immediately put it on a dirty list in the unattached state, as there is no owner whose dirty list could be used.

### PR DESCRIPTION
**Pull request by krisgiesing**
Fixes #2855

### CODE REVIEW

File: packages/flutter/lib/src/rendering/object.dart [hunk 1]
```diff
@@ -1319,6 +1340,8 @@ abstract class RenderObject extends AbstractNode implements HitTestTarget {
     return _layer;
   }
 
+  /// Whether this render object requires an update to its compositing state.
+  bool get needsCompositingBitsUpdate => _needsCompositingBitsUpdate;
```
Comment (abarth) [inline packages/flutter/lib/src/rendering/object.dart]: It looks like you're exposing this bit just for testing.  The problem with exposing these bits is that people will use them to write wrong code.  Is there a way of testing whether the `attach` code works correctly without reading these bits?  For example, presumably if we fail to dirty things properly we'll compute the wrong layer tree in the end.

File: packages/flutter/lib/src/rendering/object.dart [hunk 2]
```diff
@@ -1556,6 +1577,8 @@ abstract class RenderObject extends AbstractNode implements HitTestTarget {
   /// object, for accessibility purposes.
   Rect get semanticBounds;
 
+  /// Whether this render object requires an update to its semantics description.
+  bool get needsSemanticsUpdate => _needsSemanticsUpdate;
```
Comment (abarth) [inline packages/flutter/lib/src/rendering/object.dart]: Same comment here.

**Comment by abarth** [comment]
Generally looks good, but I wonder if there's a way to write these tests without exposing these bug-magnate bits.

File: packages/flutter/lib/src/rendering/object.dart (hunk 1 above)
Comment (krisgiesing) [inline packages/flutter/lib/src/rendering/object.dart]: markNeedsCompositingBitsUpdate() was already public; it seemed reasonable to me to add a getter for state that you can already set.  But if you're bothered by this I can look into a different mechanism. There is already an example of using a semantics listener to validate state in a different test, it's just more verbose.

Comment (abarth) [inline packages/flutter/lib/src/rendering/object.dart]: The problem is that if you let folks read these bits, they'll write code that's conditional on these bits, which is almost never correct.  I can show you lots of code in Chrome that used to read these sorts of bits and do all manner of bogus things.  :)

Comment (krisgiesing) [inline packages/flutter/lib/src/rendering/object.dart]: OK. Should the same argument apply to needsPaint and needsLayout?

Comment (abarth) [inline packages/flutter/lib/src/rendering/object.dart]: Yes, definitely.

Comment (abarth) [inline packages/flutter/lib/src/rendering/object.dart]: needsPaint looks easy to remove.  Folks (including me) are already using needsLayout for some dubious things...

File: packages/flutter/lib/src/rendering/object.dart [hunk 3]
```diff
@@ -949,6 +946,30 @@ abstract class RenderObject extends AbstractNode implements HitTestTarget {
   @override
   void attach(PipelineOwner owner) {
     super.attach(owner);
+    // If the node was dirtied in some way while unattached, make sure to add
+    // it to the appropriate dirty list now that an owner is available
+    if (_needsLayout && _relayoutSubtreeRoot != null) {
+      // Note, don't enter this block if we've never laid out at all;
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/object.dart]: the "Note" is superfluous

Comment (Hixie) [inline packages/flutter/lib/src/rendering/object.dart]: (also other cases)

File: packages/flutter/test/rendering/rendering_tester.dart [hunk 4]
```diff
@@ -51,6 +52,12 @@ class TestRenderingFlutterBinding extends BindingBase with Scheduler, Services,
     if (phase == EnginePhase.paint)
       return;
     renderView.compositeFrame();
+    if (phase == EnginePhase.composite)
```
Comment (Hixie) [inline packages/flutter/test/rendering/rendering_tester.dart]: widget_tester might also need updating

Comment (krisgiesing) [inline packages/flutter/test/rendering/rendering_tester.dart]: Ah, widget_tester had a slightly different set of phases, I'll reconcile them. Thanks.

Comment (krisgiesing) [inline packages/flutter/test/rendering/rendering_tester.dart]: Updated

Comment (Hixie) [inline packages/flutter/test/rendering/rendering_tester.dart]: I don't see the update on github?

Comment (krisgiesing) [inline packages/flutter/test/rendering/rendering_tester.dart]: I reconciled by making this class define the same EnginePhase enumeration as widget_tester.dart. The beginFrame implementations should be consistent now. The widget_tester.dart implementation already knew about semantics, but it had a slightly different definition.

**Comment by krisgiesing** [comment]
Ping?

File: packages/flutter/test/rendering/rendering_tester.dart (hunk 4 above)
Comment (Hixie) [inline packages/flutter/test/rendering/rendering_tester.dart]: Ah, ok.

LGTM.

I expected it to literally be the same enum, as opposed to two identical enums, but this is fine.

Comment (krisgiesing) [inline packages/flutter/test/rendering/rendering_tester.dart]: Yeah, there is some duplication of code between the rendering tests and the widget tests. I can file a bug to consolidate it.

**Pull request merged**